### PR TITLE
Sync fileio

### PR DESCRIPTION
**Pull request by Rinzii**
This is only the sync fileio. I plan to implement async io at some point but this can get us started for the time being. I will note I've not really tested this code super extensively.

This PR is just to bring in sync io I'll keep the branch up though for async.

### CODE REVIEW

**Comment by Rinzii** [comment]
Also a note to be made. This implementation kinda only handles binary files. I can add specializations for text files but in all honesty I don't think text files are going to be AS important till much later.

**Comment by Rinzii** [comment]
Also added a bunch of useful stuff we will need later. Not really related to this PR or branch, but just gonna bunch em' together.

File: engine/src/io/file.cpp
[the remark this answers is not in the document]
Comment (Kouros26) [reply engine/src/io/file.cpp:1]: you missed some consts on the ofstream, the streamsizes, the streampos and the filesystem functions

**Review by Kouros26 — CHANGES_REQUESTED** [review]

File: engine/src/io/file.cpp
[the remark this answers is not in the document]
Comment (Rinzii) [reply engine/src/io/file.cpp:1]: I should have got everything, but if you see anything else, lmk. ^^

**Review by Kouros26 — APPROVED** [review]

**Comment by Rinzii** [comment]
async-fileio is still a WIP branch

**Pull request merged**